src/spatial/hop_rates.jl

```diff
@@ -168,7 +172,84 @@ update rates of single species at site
 function update_hop_rate!(hop_rates::HopRatesGeneral, species, u, site, spatial_system)
     rates = hop_rates.rates
     @inbounds old_rate = rates[species, site]
-    rates[species, site] = u[species, site] * hop_rates.hop_const_cumulative_sums[species, site][end]
+    @inbounds rates[species, site] = u[species, site] * hop_rates.hop_const_cumulative_sums[species, site][end]
+    @inbounds hop_rates.sum_rates[site] += rates[species, site] - old_rate
+    old_rate
+end
+
+#################  HopRatesGeneralGrid  ######################
+"""
+Analogue of HopRatesGeneral but for CartesianGrid
+"""
+struct HopRatesGeneralGrid{F} <: AbstractHopRates
+    "hop_const_cumulative_sums[:,s,i] is the vector of cumulative sums of hopping constants of species s at site i"
+    hop_const_cumulative_sums::Array{F, 3}
+
+    "rates[s,i] is the total hopping rate of species s at site i"
+    rates::Matrix{F}
+
+    "sum_rates[i] is the total hopping rate out of site i"
+    sum_rates::Vector{F}
+end
+
+function Base.show(io::IO, ::MIME"text/plain", hop_rates::HopRatesGeneralGrid)
+    num_specs, num_sites = size(hop_rates.rates)
+    println(io, "HopRates with $num_specs species and $num_sites sites, optimized for CartesianGrid. \nHopping constants of form L_{s,i,j} where s is species, i is source and j is destination.")
+end
+
+"""
+    HopRatesGeneralGrid(hopping_constants::Matrix{Vector{F}}) where F <: Number
+
+initializes HopRates with zero rates
+"""
+function HopRatesGeneralGrid(hopping_constants::Array{F, 3}) where F <: Number
+    hop_const_cumulative_sums = mapslices(cumsum, hopping_constants, dims = 1)
+    rates = zeros(F, size(hopping_constants)[2:3])
+    sum_rates = vec(sum(rates, dims=1))
+    HopRatesGeneralGrid{F}(hop_const_cumulative_sums, rates, sum_rates)
+end
+
+function HopRatesGeneralGrid(hopping_constants::Matrix{Vector{F}}, grid) where F <: Number
+    new_hopping_constants = Array{F, 3}(undef, 2*dimension(grid), size(hopping_constants)...)
+    for ci in CartesianIndices(hopping_constants)
+        species, site = Tuple(ci)
+        new_hopping_constants[:, species, site] = pad_hop_vec(grid, site, hopping_constants[ci])
```

Comment:
`pad_hop_vec!(@view new_hopping_constant[:,species,site], grid, site,...)` ?

Reply:
Can we do the `cumsum` in place here before calling the one-argument version to save memory?

Reply:
In fact, since this is all internal why make a one-argument version? Just finish the code off inside this version...

Reply:
I put `cumsum!` in here to avoid the extra allocation. I think having two versions is convenient, e.g. for testing in `test_utils.jl`. We can also expose the main constructor to the user because it might be more convenient to pass in `Array{F, 3}` instead of `Matrix{Vector{F}}` for `hopping_constants`.